util/expensivequery/expensivequery.go

```diff
@@ -73,12 +85,214 @@ func (eqh *Handle) Run() {
+			if eqh.record.err == nil {
+				eqh.oomKillerAlert()
+			}
 		case <-eqh.exitCh:
 			return
 		}
 	}
 }
 
+type memoryUsageAlarmRecord struct {
+	err               error
+	useInstanceMemory bool // useInstanceMemory indicate whether use instance memory or system memory to alarm oom risk.
+	serverMemoryQuota uint64
+	lastRecordTime    time.Time
+
+	tmpDir              string
+	lastLogFileName     []string
+	lastProfileFileName [][]string // heap, goroutine
+}
+
+func (eqh *Handle) initMemoryUsageAlarmRecord() {
+	if config.GetGlobalConfig().Performance.ServerMemoryQuota != 0 {
+		eqh.record.serverMemoryQuota = config.GetGlobalConfig().Performance.ServerMemoryQuota
+		eqh.record.useInstanceMemory = true
+	} else {
+		eqh.record.serverMemoryQuota, eqh.record.err = memory.MemTotal()
+		if eqh.record.err != nil {
+			logutil.BgLogger().Error("Get system total memory fail.", zap.Error(eqh.record.err))
+			return
+		}
+		eqh.record.useInstanceMemory = false
+	}
+	eqh.record.lastRecordTime = time.Time{}
+	eqh.record.tmpDir = config.GetGlobalConfig().TempStoragePath
+	if alert := config.GetGlobalConfig().Performance.MemoryUsageAlarmRatio; alert == 0 || alert == 1 {
+		eqh.record.err = errors.New("close memory usage alarm recorder")
+	}
+	eqh.record.lastProfileFileName = make([][]string, 2)
+}
+
+// If Performance.ServerMemoryQuota is set, use instance memory usage and ServerMemoryQuota * 80% to check oom risk.
+// If Performance.ServerMemoryQuota is not set, use system memory usage and total memory * 80% to check oom risk.
+func (eqh *Handle) oomKillerAlert() {
+	var memoryUsage uint64
+	instanceStats := &runtime.MemStats{}
+	if eqh.record.useInstanceMemory {
+		runtime.ReadMemStats(instanceStats)
+		memoryUsage = instanceStats.HeapAlloc
+	} else {
+		memoryUsage, eqh.record.err = memory.MemUsed()
+		if eqh.record.err != nil {
+			logutil.BgLogger().Error("Get system usage memory fail.", zap.Error(eqh.record.err))
+			return
+		}
+	}
+
+	// If we use instance memory usage to check oom risk, we also need use NextGC compared with the quota.
+	// Maybe the instance will oom but the HeapAlloc isn't updated after golang GC.
+	// Go's GC don't take the system memory limit into consideration, that means:
+	//
+	// 1. You have 16G physical memory;
+	// 2. You have 9G live objects;
+	// 3. Go will run GC when memory usage reaches 9G * (1 + GOGC / 100) = 18G, OOM.
+	if float64(memoryUsage) > float64(eqh.record.serverMemoryQuota)*config.GetGlobalConfig().Performance.MemoryUsageAlarmRatio ||
+		(eqh.record.useInstanceMemory && instanceStats.NextGC > eqh.record.serverMemoryQuota) {
+		// At least ten seconds between two recordings that memory usage is less than threshold (default 80% system memory).
+		// If the memory is still exceeded, only records once.
+		if time.Since(eqh.record.lastRecordTime) > 10*time.Second {
+			eqh.oomRecord(memoryUsage)
+		}
+		eqh.record.lastRecordTime = time.Now()
+	}
+}
+
+func (eqh *Handle) oomRecord(memUsage uint64) {
```

Comment:
1. It'll be better to define this as a member method for memUsageAlarm?
2. oomRecord --> doRecord ?